Soundeffect editor accessibility

glastonbridge opened the pull request:
[https://soundeffect-editor-accessibi.review-pxt.pages.dev](https://soundeffect-editor-accessibi.review-pxt.pages.dev/)

### Summary

Overall accessibility improvements for keyboard navigation including visual feedback. Changes include
* Navigating the gallery in a Grid pattern [as recommended by WAI-ARIA](https://www.w3.org/WAI/ARIA/apg/patterns/grid/)
* Improved combo box behaviour
* Highlighting the selected waveform is now more visible
* Keeps the cursor in a focus trap inside the Sound Effect editor until escape is pressed or the window is otherwise closed
* Keyboard controls for the frequency and volume sliders
* Fixed the browser console error relating to Radio Groups

Note that this changes the Selected Waveform Radio Group for all users. This is because other recent theming changes have reduced the contrast between the active button and other radio buttons, and this makes it strongly visible again. Ie.:

**In master**

<img width="163" alt="image" src="https://github.com/user-attachments/assets/0665bce1-366c-4185-86ba-a8a2bdfc9267" />

**In this branch**

<img width="173" alt="image" src="https://github.com/user-attachments/assets/4e18c6ec-7bf6-4056-9870-ac45c2e65ad0" />

### Walkthrough video using tab, arrow keys, space, enter, escape

https://github.com/user-attachments/assets/0d04efb9-a86d-49ea-90ef-a3efe8d791bc

### To test

1. Go to the Live demo above
2. Create a project containing a Sound Effect
3. Enable Accessible Blocks in the cog menu 
4. Navigate to the Sound Effect using the arrow keys and press enter
5. Explore the UI

microbit-robert commented:
This is so much better than it was! Found a couple of minor issues, a suggested improvement and some nitpicks. I also wonder if they'll revert back to the current styling when they do more theme fixes, but your version is arguably clearer.

Issues:
- Hitting 'Tab' inexplicably moves the focus to the 'Editor' tab toggle when a waveform that **isn't the selected waveform** has focus (can screen record if this isn't clear)
- Hitting 'Tab' inexplicably moves the focus to the 'Editor' tab toggle when the duration dropdown is open and **the first item isn't the item with focus**
- Spacebar inside the frequency / volume mini editor plays the sound, but if your focus is the on the Effect or Interpolation dropdowns, 'Space' also either opens the dropdown or selects an option depending on the state. Perhaps the sound shouldn't be played if the dropdowns have focus?

Improvement:
- When inside the frequency / volume mini editor, pressing 'Space' plays the sound. 'Enter' should probably do the same as 'Space' after the behaviour conflicts above have been resolved.

Nitpicks:
- The focus is initially on the entire div, and the focus-visible outline only appears around half of the div because it's hidden by the header and the waveform. Should focus be on the 'Editor' toggle when the div is opened instead? I think this more closely matches the behaviour of other Blockly field editors and resolves the focus outline issue.
- The focus-visible outline around the play button falls off the right-hand side of the div. Is it worth pulling that button in a bit?

glastonbridge commented:
Hey @microbit-robert, I've had a go at addressing those issues, let me know what you think!

microbit-robert commented on react-common/components/controls/Dropdown.tsx:
```diff
@@ -31,20 +31,23 @@ export const Dropdown = (props: DropdownProps) => {
     const [ expanded, setExpanded ] = React.useState(false);
 
-    let container: HTMLDivElement;
+    const container = React.useRef<HTMLDivElement>();
+    const dropdownButton = React.useRef<HTMLButtonElement>();
```
These should start off as `null` with the type `<HTML... | null>`

microbit-robert commented on webapp/src/components/soundEffectEditor/SoundGallery.tsx:
```diff
@@ -18,33 +18,109 @@ export interface SoundGalleryProps {
+        switch(event.code) {
+            case "ArrowDown":
+                selectItemRefs.current[next].focus();
+                console.log("focused ",selectItemRefs.current, next);
```
Remove debug

microbit-robert commented on webapp/src/components/soundEffectEditor/SoundGallery.tsx:
```diff
@@ -18,33 +18,109 @@ export interface SoundGalleryProps {
+        next: number,
+        current: number,
+        event: React.KeyboardEvent<HTMLElement>) {
+            console.log("Selectnav", event.code);
```
Remove debug

microbit-robert commented on webapp/src/components/soundEffectEditor/SoundGallery.tsx:
```diff
@@ -18,33 +18,109 @@ export interface SoundGalleryProps {
-
 const SoundGalleryEntry = (props: SoundGalleryItemProps) => {
-    const { sound, name, useMixerSynthesizer } = props;
+    const { 
```
Remove trailing space
```suggestion
    const {
```

microbit-robert commented on webapp/src/components/soundEffectEditor/SoundGallery.tsx:
```diff
@@ -18,33 +18,109 @@ export interface SoundGalleryProps {
+    const selectItemRefs = React.useRef<Record<string,HTMLElement>>({});
+    const playItemRefs = React.useRef<Record<string,HTMLElement>>({});
+
+    function selectNav(
```
A bit unusual to see this style of function declaration in a React component.

kmcnaught commented:
On 3, I note that @microbit-robert gave the icon grid `Home/End` functionality, which I approved of. But actually it works on a row basis, where End takes you to the end of the current row rather than the bottom row. Which wouldn't be particularly helpful here. And in both cases PgUp/PgDn work to scroll, without moving focus. 

I have previously asked for Page Up/Down to move focus in the toolbox, I think the same pattern would be helpful here in the grid, when you're zoomed in enough to have a few pages to navigate through. 

Might be worth a wider discussion of what Home/End/PgUp/PgDn should do in this and other situations, so we can find a consistent pattern.

Having played with zoom more, I have a couple more niggles:
- The wrap from top to bottom of grid is confusing me. I think I would prefer hard stops. 
- I do keep finding myself pressing `Down` from try to get to the grid after selecting the Gallery. It feels like adding this wouldn't be particularly standard behaviour, but would perhaps feel natural given all the other arrow nav that we are getting people accustomed to. Anyone else have opinions on this?

kmcnaught commented:
Oh look, here's a different version of the toggle pattern.  I don't mind this one behaving differently to the code editor toggle, since the focus outline implies the arrow nav. Perhaps copy the pattern from the melody sandbox for the sound effect editor?

![melody sandbox toggle with outline over both options](https://github.com/user-attachments/assets/3a86d5d2-6f41-46a4-8e91-e50faadc6ee3)

microbit-robert commented:
> On 3, I note that @microbit-robert gave the icon grid `Home/End` functionality, which I approved of. But actually it works on a row basis, where End takes you to the end of the current row rather than the bottom row. Which wouldn't be particularly helpful here. And in both cases PgUp/PgDn work to scroll, without moving focus.

We aligned with Blockly's approach to their grid and colour plugins for this. Ctrl-Home and Ctrl-End work to get the very first and very last element. If we want to move to PgUp/PgDn then we need communicate this to Blockly. We can't have it working in different ways for different blocks.

microbit-robert commented:
> Oh look, here's a different version of the toggle pattern. I don't mind this one behaving differently to the code editor toggle, since the focus outline implies the arrow nav. Perhaps copy the pattern from the melody sandbox for the sound effect editor?

I think we should align these toggles where it makes sense. The main Blocks / JavaScript one feels different to the others. It looks like bit like a tab toggle, but that's not really its behaviour. We also have another button attached to it which opens a dropdown.

glastonbridge commented:
Thank you for all your feedback @microbit-robert @kmcnaught ! I've addressed all the code tidying issues and found a couple of other bugs that are now dealt with. In terms of UI, the things that are in this:

* The gallery is now a single tab stop
* Home and End work
* Arrow nav in the gallery does not wrap
* Rebased against master now that the blockly key handler PR is merged https://github.com/microsoft/pxt/pull/10554

Stuff that's not in here:

* I will raise an issue to make the toggle titlebars work consistently
* I reverted the waveform contrast highlighting, and raised https://github.com/microsoft/pxt/issues/10563 instead
* I did not add pageup/pagedown as there seems to be some question over that.

Let me know if this all seems okay

glastonbridge commented:
Update: issue for the title toggle inconsistency https://github.com/microbit-foundation/blockly-project/issues/210

microbit-robert commented:
I made some tweaks to clean up some code and fix a bug (the sound effect editor `useEffect` has no deps so we were adding an event listener for every key press). I think this is now good to raise as a PR in `pxt`.

glastonbridge commented:
Just dropping in because I happened to be online, thanks for taking this forward!

> the sound effect editor useEffect has no deps so we were adding an event listener for every key press

This now causes the old sound to get cached. To replicate, try using the preview sound in the gallery, and note that it plays the preview sound and the existing selected sound at the same time. I think the old `play` function is hanging around with an older sound attached.

Off the top of my head, a possible fix is to create a useCallback for `play` (with the correct dependencies) and then use play as a dependency for the useEffect.

microbit-robert commented:
> Just dropping in because I happened to be online, thanks for taking this forward!
> 
> > the sound effect editor useEffect has no deps so we were adding an event listener for every key press
> 
> This now causes the old sound to get cached. To replicate, try using the preview sound in the gallery, and note that it plays the preview sound and the existing selected sound at the same time. I think the old `play` function is hanging around with an older sound attached.
> 
> Off the top of my head, a possible fix is to create a useCallback for `play` (with the correct dependencies) and then use play as a dependency for the useEffect.

Good catch! Yes, I think that's the right fix. Will address shortly.

microbit-robert commented on webapp/src/components/soundEffectEditor/SoundEffectEditor.tsx:
```diff
@@ -84,7 +77,36 @@ export const SoundEffectEditor = (props: SoundEffectEditorProps) => {
+
+    React.useEffect(() => {
+        const keyListener = (ev: KeyboardEvent) => {
+            // Ignore all keys that could be used for accessibility navigation
+            // Enter is exempt to maintain the same behaviour with Space (" ")
+            if ((ev.key.length !== 1 && ev.key !== "Enter") || ev.metaKey || ev.ctrlKey || /[0-9]/.test(ev.key)) return;
```
Move to using a React keydown handler [acdbbae](https://github.com/microbit-matt-hillsdon/pxt/pull/46/commits/acdbbae3ea758a67a71ed844097b69a96be653c8)

I'll draw attention to this when we raise the PR. At the moment, if the field editor is open, and focus is anywhere else in the document, button presses still cause the sound to play.

microbit-robert commented:
Closing as this is now raised in `pxt`
